File: September 2026/2026-09-06/396-rotate-function.py
# 396-rotate-function.cpp;dp;help;leetcode

class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        n = len(nums)
        total_sum = sum(nums)
        prev = sum(nums[i] * i for i in range(n))
        ans = prev
        index = n - 1

        for _ in range(1, n):
            prev = prev + total_sum - n * nums[index]

            index -= 1

            ans = max(ans, prev)

        return ans

File: September/2026-09-06/396/396-rotate-function.cpp
// 396-rotate-function.cpp;dp;help;leetcode

class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int n = nums.size();
        int sum = 0, prev = 0;
        for(int i = 0; i < n; i++) {
            prev += nums[i] * i;
            sum += nums[i];
        }

        int index = n - 1;
        int ans = prev;
        for(int i = 1; i < n; i++) {
            int x = sum-nums[index] - (n - 1) * nums[index];
            prev = prev + x;
            index--;
            ans = max(ans, prev);
        }

        return ans;
    }
};